lib/connection/msg.js

```diff
@@ -0,0 +1,293 @@
+'use strict';
+
+// Implementation of OP_MSG spec:
+// https://github.com/mongodb/specifications/blob/master/source/message/OP_MSG.rst
+//
+// struct Section {
+//   uint8 payloadType;
+//   union payload {
+//       document  document; // payloadType == 0
+//       struct sequence { // payloadType == 1
+//           int32      size;
+//           cstring    identifier;
+//           document*  documents;
+//       };
+//   };
+// };
+
+// struct OP_MSG {
+//   struct MsgHeader {
+//       int32  messageLength;
+//       int32  requestID;
+//       int32  responseTo;
+//       int32  opCode = 2013;
+//   };
+//   uint32      flagBits;
+//   Section+    sections;
+//   [uint32     checksum;]
+// };
+
+const opcodes = require('../wireprotocol/shared').opcodes;
+const databaseNamespace = require('../wireprotocol/shared').databaseNamespace;
+
+// Incrementing request id
+let _requestId = 0;
+
+// Msg Flags
+const OPTS_CHECKSUM_PRESENT = 1;
+const OPTS_MORE_TO_COME = 2;
+const OPTS_EXHAUST_ALLOWED = 1 >> 16;
+
+class Msg {
+  constructor(bson, ns, command, options) {
+    // Basic options needed to be passed in
+    if (command == null) throw new Error('query must be specified for query');
+
+    // Basic options
+    this.bson = bson;
+    this.ns = ns;
+    this.command = command;
+    this.command.$db = databaseNamespace(ns);
+
+    // Ensure empty options
+    this.options = options || {};
+
+    // Additional options
+    this.requestId = Msg.getRequestId();
+
+    // Serialization option
+    this.serializeFunctions =
```

Comment:
Can we document these options?

Reply:
should we document these everywhere? They are just options eventually passed down to BSON for deserialization. I think we've adequately documented them at the top level, and don't need to see these docs at every step they are considered.

Reply:
That's fair. My thought process was that if we need to do anything similar to options management again, it's easiest to see where options are used by scanning through the documentation. That being said, I only care 1/10 on this so I'm fine with not documenting these.